qutip/core/data/csc.pyx

```diff
@@ -0,0 +1,434 @@
+    cdef base.idxint i
+    for i in range(dimension):
+        out.data[i] = scale
+        out.row_index[i] = i
+        out.col_index[i] = i
+    out.col_index[dimension] = dimension
+    return out
+
+
+cpdef CSC fast_from_scipy(object sci):
+    """
+    Fast path construction from scipy.sparse.csc_matrix.  This does _no_ type
+    checking on any of the inputs, and should consequently be considered very
+    unsafe. This is primarily for use in the unpickling operation.
+    """
+    cdef CSC out = CSC.__new__(CSC)
+    out.shape = sci.shape
+    out._deallocate = False
+    out._scipy = sci
+    out.data = <double complex *> cnp.PyArray_GETPTR1(sci.data, 0)
+    out.row_index = <base.idxint *> cnp.PyArray_GETPTR1(sci.indices, 0)
+    out.col_index = <base.idxint *> cnp.PyArray_GETPTR1(sci.indptr, 0)
+    out.size = cnp.PyArray_SIZE(sci.data)
+    return out
+
+
+cpdef CSC from_csr(CSR matrix):
+    """Transform a CSR to CSC."""
+    cdef CSR transposed = transpose_csr(matrix)
+    return from_tr_csr(transposed, False)
+
+
+cpdef CSC from_dense(Dense matrix):
+    # Assume worst-case scenario for non-zero.
+    cdef CSC out = empty(matrix.shape[0], matrix.shape[1],
+                         matrix.shape[0]*matrix.shape[1])
+    cdef size_t row, col, ptr_in, ptr_out=0, row_stride, col_stride
+    row_stride = 1 if matrix.fortran else matrix.shape[1]
+    col_stride = matrix.shape[0] if matrix.fortran else 1
+    out.col_index[0] = 0
+    for col in range(matrix.shape[1]):
+        ptr_in = col_stride * col
+        for row in range(matrix.shape[0]):
+            if matrix.data[ptr_in] != 0:
+                out.data[ptr_out] = matrix.data[ptr_in]
+                out.row_index[ptr_out] = row
+                ptr_out += 1
+            ptr_in += row_stride
+        out.col_index[col + 1] = ptr_out
+    return out
+
+
+cpdef CSR as_tr_csr(CSC matrix, bint copy=True):
+    """ Return a CSR which is the transposed to this CSC.
+    If copy is False, this CSR is a view on the CSC with no data ownership.
+    """
+    cdef CSR out = CSR.__new__(CSR)
+    out.data = matrix.data
+    out.col_index = matrix.row_index
+    out.row_index = matrix.col_index
+    out.size = matrix.size
+    out._deallocate = False
+    out.shape = (matrix.shape[1], matrix.shape[0])
+    if copy:
+        return out.copy()
+    else:
+        return out
+
+
+cpdef CSC from_tr_csr(CSR matrix, bint copy=True):
+    """ Return a CSC which is the transposed to this CSR.
+    If copy is False, steal data ownership from the CSR.
+    """
+    cdef CSC out = CSC.__new__(CSC)
+    out.data = matrix.data
+    out.col_index = matrix.row_index
+    out.row_index = matrix.col_index
+    out.size = matrix.size
+    out.shape = (matrix.shape[1], matrix.shape[0])
+    if copy:
+        out._deallocate = False
+        return out.copy()
+    else:
+        out._deallocate = True
+        matrix._deallocate = False
+        return out
```

Comment:
This looks like it'll segfault - it's very difficult to manage this sort of data ownership like this, which is why I deferred to `numpy` as soon as it got even a little complicated.  In this case, there's a bug if `matrix._deallocate` is `False` on entry to the function and `copy=False`, then `out._deallocate` will get set to `True`, but that isn't safe since `matrix` doesn't own its data.  You also _can't_ fix this with

```python
else:
    out._deallocate = matrix._deallocate
```

because if `out` goes out of scope before `matrix`, then access to `matrix` will segfault.  We can't allow any segfaulting operation to be callable from Python space (and even leaving it in C space is very questionable).

I can think of two ways to make this safe, but I'm not convinced by either of them:

1. add a reference to the underlying object into the returned object to ensure they never go out of scope together, but this is basically reinventing Python and numpy's garbage collection and ownership semantics, and I have no doubts that we'll get it wrong 100 times.
2. force the creation of the `scipy` view of the input, and then use our fast `scipy.sparse.csr_matrix`/`...csc_matrix` constructors to produce one for the output and then fake an unpickle from that.  That lets numpy handle the data ownership safely, but it'll be difficult to follow the internal logic.

If I had to pick, I'd go with option 2, but I'm not totally certain that's the best way.

Reply:
`as_tr_csr` and `from_tr_csr` are to be used in pair when using CSR specializations for CSC. So instead of needing them to work in all cased we can have them fail when miss-used to signal us to rework the affected specializations.
I added a check in `to_scipy` to refuse to create scipy matrix if `_scipy` does not exist it does not own data and `from_tr_csr` will refuse to steal ownership if the target does not have it.

I made them generic ValueError, maybe BufferError is more appropriate...